core/lib/object_store/src/s3.rs

```diff
@@ -0,0 +1,225 @@
+        if let Some(endpoint) = endpoint.clone() {
+            tracing::info!(endpoint=%endpoint, "using S3 endpoint defined in storage config");
+            sdk_config = sdk_config.endpoint_url(endpoint);
+        }
+        let sdk_config = sdk_config.load().await;
+        let client = Client::new(&sdk_config);
+
+        Ok(Self {
+            endpoint: endpoint.unwrap_or_default(),
+            bucket_prefix,
+            client,
+        })
+    }
+
+    fn get_client_config(auth_mode: S3StoreAuthMode) -> ConfigLoader {
+        match auth_mode {
+            S3StoreAuthMode::AuthenticatedWithCredentialFile(path) => {
+                let profile_files = EnvConfigFiles::builder()
+                    .with_file(EnvConfigFileKind::Credentials, path)
+                    .build();
+                aws_config::defaults(BehaviorVersion::latest()).profile_files(profile_files)
+            }
+            S3StoreAuthMode::Anonymous => {
+                aws_config::defaults(BehaviorVersion::latest()).no_credentials()
+            }
+        }
+    }
+
+    fn filename(bucket: &str, filename: &str) -> String {
+        format!("{bucket}/{filename}")
+    }
+}
+
+impl From<ByteStreamError> for ObjectStoreError {
+    fn from(err: ByteStreamError) -> Self {
+        ObjectStoreError::Other {
+            source: err.into(),
+            is_retriable: true,
+        }
+    }
+}
+
+impl From<anyhow::Error> for ObjectStoreError {
+    fn from(err: anyhow::Error) -> Self {
+        ObjectStoreError::Initialization {
+            source: err.into(),
+            is_retriable: false,
+        }
+    }
+}
+
+impl<T> From<SdkError<T>> for ObjectStoreError
+where
+    T: std::error::Error + Send + Sync + 'static,
+{
+    fn from(err: SdkError<T>) -> Self {
+        match &err {
+            SdkError::ConstructionFailure(_) => ObjectStoreError::Initialization {
+                source: err.into(),
+                is_retriable: false,
+            },
+            SdkError::DispatchFailure(_) | SdkError::TimeoutError(_) => ObjectStoreError::Other {
+                source: err.into(),
+                is_retriable: true,
+            },
+            SdkError::ResponseError(http_response) => {
+                let code = http_response.raw().status().as_u16();
+                if code == StatusCode::NOT_FOUND.as_u16() {
+                    ObjectStoreError::KeyNotFound(err.into())
+                } else if code == StatusCode::FORBIDDEN.as_u16() {
+                    ObjectStoreError::Initialization {
+                        source: err.into(),
+                        is_retriable: false,
+                    }
+                } else {
+                    ObjectStoreError::Other {
+                        source: err.into(),
+                        is_retriable: true,
+                    }
+                }
+            }
+            SdkError::ServiceError(http_response) => {
+                let code = http_response.raw().status().as_u16();
```

Comment:
Nit: It looks like handling of this and the previous variant can be extracted to a function in order to avoid code duplication.

Reply:
I tried to do that, but unfortunately `http_response.raw()` returns private Response type, so I can't create a function which accept it. And `http_response` type is different in both cases.